### PR TITLE
Es item -- Fourfront. NOT READY

### PR DESCRIPTION
**Pull request by carlvitzthum**
**DO NOT FORGET TO UPDATE SNOVAULT BRANCH IN BUILDOUT.CFG**

This branch is the Fourfront complement to https://github.com/4dn-dcic/snovault/pull/111. It adds some functionality specific to the new Elasticsearch items and a few other misc things.

**Summary**
- dev_servers.py changed to run `create_mapping` before and after loading test data. This is done to initialize any ES-item indices before loading, as well as deduplicate the queue before serving the application (and thus indexing)
- Refactored `User._update` to make more sense. Now does not does gets to multiple labs to get display titles. This code should be fully deprecated in the future for a more full-featured subscriptions
- `/counts` refactored a bit. May be slightly less performant, but is now much more maintainable
- Refactored some indexing-related test features. This previously in `.features.conftest` have been moved to `workbook_fixtures`
- Removed last reference to `linkFrom` in the testing types
- Other random test fix

### CODE REVIEW

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 8
- Added 3
           

Complexity increasing per file
==============================
- src/encoded/dev_servers.py  1
- src/encoded/tests/workbook_fixtures.py  2
         

Complexity decreasing per file
==============================
+ src/encoded/types/user.py  -3
+ src/encoded/root.py  -9
+ src/encoded/tests/test_types_user.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692436&bid=15687384)

File: src/encoded/tests/test_indexing.py [hunk 1]
```diff
@@ -8,10 +8,12 @@
 import time
 import os
 from encoded.verifier import verify_item
+from snovault import DBSESSION, TYPES
+from snovault.elasticsearch import create_mapping, ELASTIC_SEARCH
 from snovault.elasticsearch.interfaces import INDEXER_QUEUE
 from snovault.elasticsearch.indexer_utils import get_namespaced_index
 from elasticsearch.exceptions import NotFoundError
-from .features.conftest import app_settings, app as conf_app
+from .workbook_fixtures import app_settings
```
Comment (j1z0) [inline src/encoded/tests/test_indexing.py]: Codacy Issue found: ['.workbook_fixtures.app_settings' imported but unused](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692436)

File: src/encoded/tests/test_indexing.py [hunk 2]
```diff
@@ -20,26 +22,38 @@
...
 
 
-@pytest.yield_fixture(scope='session')
-def app(app_settings, use_collections=TEST_COLLECTIONS):
-    """
-    Use to pass kwargs for create_mapping to conftest app
-    """
-    for app in conf_app(app_settings, collections=use_collections, skip_indexing=True):
-        yield app
+@pytest.yield_fixture(scope='session', params=[False])
```
Comment (j1z0) [inline src/encoded/tests/test_indexing.py]: Codacy Issue found: [redefinition of unused 'app_settings' from line 16](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692436)

File: src/encoded/tests/test_indexing.py (hunk 1 above)
Comment (j1z0) [inline src/encoded/tests/test_indexing.py]: Codacy Issue found: [Unused app_settings imported from workbook_fixtures](https://app.codacy.com/app/4dn/fourfront/pullRequest?prid=4692436)

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 8
- Added 5
           

Complexity increasing per file
==============================
- src/encoded/tests/workbook_fixtures.py  2
- src/encoded/dev_servers.py  1
         

Complexity decreasing per file
==============================
+ src/encoded/types/user.py  -3
+ src/encoded/root.py  -9
+ src/encoded/tests/test_types_user.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=4692436&bid=15687384)

File: src/encoded/tests/test_indexing.py (hunk 1 above)
Comment (j1z0) [inline src/encoded/tests/test_indexing.py]: Codacy Issue found: ['.workbook_fixtures.app_settings' imported but unused](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=4692436)

File: src/encoded/tests/test_types_init_collections.py [hunk 3]
```diff
@@ -233,6 +234,7 @@ def test_tracking_item_display_title_download(download_tracking):
 
 
 def test_tracking_item_display_title_other(jupyterhub_session):
+    from datetime import datetime
```
Comment (j1z0) [inline src/encoded/tests/test_types_init_collections.py]: Codacy Issue found: [Unused variable 'datetime'](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=4692436)

Comment (j1z0) [inline src/encoded/tests/test_types_init_collections.py]: Codacy Issue found: ['datetime.datetime' imported but unused](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=4692436)

File: src/encoded/tests/test_indexing.py (hunk 1 above)
Comment (j1z0) [inline src/encoded/tests/test_indexing.py]: Codacy Issue found: [Unused app_settings imported from workbook_fixtures](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=4692436)

File: src/encoded/tests/test_indexing.py (hunk 2 above)
Comment (j1z0) [inline src/encoded/tests/test_indexing.py]: Codacy Issue found: [redefinition of unused 'app_settings' from line 16](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=4692436)

File: src/encoded/tests/test_types_user.py [hunk 4]
```diff
@@ -16,51 +16,44 @@ def user_w_lab(testapp, lab):
...
+def test_subscriptions_dont_duplicate_on_update(registry, lab):
     # run _update on a new user with already formed subscriptions.
     # Ensure they're not duplicated in the process
     user_data = {
         'first_name': 'User',
         'last_name': 'McUser',
         'email': 'user@mcuser.org',
         'status': 'current',
+        'lab': lab['uuid'],
         'subscriptions': [{'url': '?lab.uuid=' + lab['uuid'] + '&sort=-date_created',
```
[the remark this answers is not in the document]
Comment (willronchetti) [reply src/encoded/tests/test_types_user.py]: `Lab` in this case I believe is the result of the post so it will always have uuid.

**Pull request merged**